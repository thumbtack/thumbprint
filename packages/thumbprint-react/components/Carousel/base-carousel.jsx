import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import rotate from 'rotate-array';
import range from 'lodash/range';
import styles from './base-carousel.module.scss';

// Utility function from https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state
function usePrevious(value) {
    const ref = useRef();
    useEffect(() => {
        ref.current = value;
    });
    return ref.current;
}

export default function BaseCarousel({
    children,
    selectedIndex,
    animationDuration,
    visibleCount,
    spacing,
}) {
    // When animating, `prevSelectedIndex` is the value of `selectedIndex` before the
    // animation began. Once the animation is complete, it becomes the same as `selectedIndex`.
    const [prevSelectedIndex, setPrevSelectedIndex] = useState(selectedIndex);
    const [isAnimating, setIsAnimating] = useState(false);
    const [isSuspensed, setIsSuspensed] = useState(false);

    const prevSelectedIndexProp = usePrevious(selectedIndex);

    function setAnimating() {
        setIsAnimating(true);
        setIsSuspensed(false);

        setTimeout(() => {
            setIsAnimating(false);
            setIsSuspensed(true);
            setPrevSelectedIndex(selectedIndex);

            // We suspend the CSS animation property for a very brief window before
            // re-enabling. This gap allows the component to re-render the new list
            // without the items "sliding" back into place. Once the new items are set up,
            // we re-enable the animation property ready for the next transition.
            setTimeout(() => {
                setIsSuspensed(false);
            }, 50);
        }, animationDuration);
    }

    function reorder(items) {
        // The `prevSelectedIndex` doesn't update until the animation is done, so we want to
        // use that ordering until the animation is complete.
        return rotate(items, -1 * Math.floor(prevSelectedIndex));
    }

    useEffect(() => {
        if (selectedIndex !== prevSelectedIndexProp) {
            setAnimating();
        }
    });

    const itemWidth = 1 / visibleCount;

    // If `selectedIndex` is `1.5`, this value is `0.5`. This is used when the consumer wants
    // to partially display children.
    const fractionalIndexOffset = Math.abs(prevSelectedIndex - Math.floor(prevSelectedIndex));

    // The `adjustedIndex`, when animating, is the number of items to slide. If the value is
    // negative it means that we should slide to the left. It gets reset to `0` when the
    // animation is complete.
    const adjustedIndex = selectedIndex - prevSelectedIndex;

    // This is equal to `fractionalIndexOffset` in the resting state but changes while the
    // transition is occuring.
    const translateX = itemWidth * (adjustedIndex + fractionalIndexOffset) * -100;

    // An array of the flex order of the items.
    const orderedChildren = reorder(range(children.length));

    // Sometimes we need to duplicate the children so that the carousel can display properly.
    // This is especially needed when animating. Imagine that a 4-item card with 3 visible
    // items wants to animate 2 items to the right. Normally, this would require 5 items to
    // animate without displaying an empty spot. Since there are only 4 items, we temporarily
    // duplicate the `children` until the animation is complete.
    const shouldRenderDuplicateChildren = isAnimating;

    return (
        <div className={styles.root}>
            <ul
                className={styles.wrapper}
                style={{
                    transform: `translateX(${translateX}%)`,
                    transition: isSuspensed ? 'none' : `transform ${animationDuration}ms ease`,
                    width: `calc(100% + ${spacing})`,
                }}
            >
                {React.Children.map(children, (child, i) => (
                    <li
                        key={i}
                        className={styles.item}
                        style={{
                            width: `${itemWidth * 100}%`,
                            paddingRight: spacing,
                            order: orderedChildren[i],
                        }}
                    >
                        {child}
                    </li>
                ))}

                {/* Temporary elements that appear to the left. */}
                {shouldRenderDuplicateChildren &&
                    React.Children.map(children, (child, i) => (
                        <li
                            key={orderedChildren.length + i}
                            className={styles.item}
                            style={{
                                width: `${itemWidth * 100}%`,
                                paddingRight: spacing,
                                order: orderedChildren[i] + orderedChildren.length,
                                transform: `translateX(${children.length * -200}%)`,
                            }}
                        >
                            {child}
                        </li>
                    ))}

                {/* Temporary elements that appear to the right. */}
                {shouldRenderDuplicateChildren &&
                    React.Children.map(children, (child, i) => (
                        <li
                            key={orderedChildren.length + i * 2}
                            className={styles.item}
                            style={{
                                width: `${itemWidth * 100}%`,
                                paddingRight: spacing,
                                order: (orderedChildren[i] + orderedChildren.length) * 2,
                                transform: `translateX(${children.length * -100}%)`,
                            }}
                        >
                            {child}
                        </li>
                    ))}
            </ul>
        </div>
    );
}

BaseCarousel.propTypes = {
    /**
     * The items that render horizontally within the carousel.
     */
    children: PropTypes.node.isRequired,
    /**
     * The index of the left-most item to display in the carousel. Supports whole numbers and
     * decimals.
     */
    selectedIndex: PropTypes.number,
    /**
     * The number of items that are visible at once.
     */
    visibleCount: PropTypes.number,
    /**
     * The amount of space separating each item.
     */
    spacing: PropTypes.string,
    /**
     * The duration of the animation (in milliseconds).
     */
    animationDuration: PropTypes.number,
};

BaseCarousel.defaultProps = {
    selectedIndex: 0,
    visibleCount: 1,
    spacing: '0px',
    animationDuration: 400,
};
