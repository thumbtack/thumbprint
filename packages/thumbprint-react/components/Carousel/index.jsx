import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import Swipeable from 'react-swipeable';
import BaseCarousel from './base-carousel.jsx';

export default function Carousel({
    children,
    spacing,
    visibleCount,
    selectedIndex,
    onSelectedIndexChange,
}) {
    const [dragIndexOffset, setDragIndexOffset] = useState(0);
    const isDragging = dragIndexOffset !== 0;
    const containerRef = useRef();

    function onDrag(deltaX) {
        const size = containerRef.current && containerRef.current.getBoundingClientRect();

        // Set the `dragIndexOffset` so that the carousel can visually move before we send
        // the new index to the parent.
        setDragIndexOffset((deltaX / size.width) * visibleCount);
    }

    function onDragEnd() {
        onSelectedIndexChange(selectedIndex + dragIndexOffset);
        setDragIndexOffset(0);
    }

    return (
        <div ref={containerRef}>
            <Swipeable
                preventDefaultTouchmoveEvent
                onSwiping={(_, deltaX) => onDrag(deltaX)}
                onSwiped={onDragEnd}
            >
                <BaseCarousel
                    selectedIndex={selectedIndex + dragIndexOffset}
                    visibleCount={visibleCount}
                    spacing={spacing}
                    animationDuration={isDragging ? 0 : undefined}
                >
                    {children}
                </BaseCarousel>
            </Swipeable>
        </div>
    );
}

Carousel.propTypes = {
    /**
     * The items in the carousel that appear horizontally.
     */
    children: PropTypes.node.isRequired,
    /**
     * The index of the left-most item to display in the carousel. Supports all numbers.
     */
    selectedIndex: PropTypes.number.isRequired,
    /**
     * This function is called while a user is dragging the carousel and receives the new index.
     * This should update the `selectedIndex` prop similarly to an `onChange` in a controlled
     * input.
     */
    onSelectedIndexChange: PropTypes.func.isRequired,
    /**
     * The number of items that are visible at once.
     */
    visibleCount: PropTypes.number,
    /**
     * The amount space separating each item. Supports CSS values such as `8px` or `0.5em`.
     */
    spacing: PropTypes.string,
};

Carousel.defaultProps = {
    visibleCount: 1,
    spacing: '0px',
};
