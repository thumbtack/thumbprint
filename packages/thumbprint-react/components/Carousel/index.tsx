import React, { useState, useRef } from 'react';
import Swipeable from 'react-swipeable';
import BaseCarousel from './base-carousel';

interface PropTypes {
    /**
     * The items in the carousel that appear horizontally.
     */
    children: React.ReactNode;
    /**
     * The index of the left-most item to display in the carousel. Supports all numbers.
     */
    selectedIndex: number;
    /**
     * This function is called while a user is dragging the carousel and receives the new index.
     * This should update the `selectedIndex` prop similarly to an `onChange` in a controlled
     * input.
     */
    onSelectedIndexChange: (newIndex: number) => void;
    /**
     * The number of items that are visible at once.
     */
    visibleCount?: number;
    /**
     * The amount space separating each item. Supports CSS values such as `8px` or `0.5em`.
     */
    spacing?: string;
}

export default function Carousel({
    children,
    spacing = '0px',
    visibleCount = 1,
    selectedIndex,
    onSelectedIndexChange,
}: PropTypes): JSX.Element {
    const [dragIndexOffset, setDragIndexOffset] = useState(0);
    const isDragging = dragIndexOffset !== 0;
    const containerRef = useRef<HTMLDivElement>(null);

    function onDrag(deltaX: number): void {
        const size = containerRef.current ? containerRef.current.getBoundingClientRect() : null;

        if (size && size.width) {
            // Set the `dragIndexOffset` so that the carousel can visually move before we send
            // the new index to the parent.
            setDragIndexOffset((deltaX / size.width) * visibleCount);
        }
    }

    function onDragEnd(): void {
        onSelectedIndexChange(selectedIndex + dragIndexOffset);
        setDragIndexOffset(0);
    }

    return (
        <div ref={containerRef}>
            <Swipeable
                preventDefaultTouchmoveEvent
                onSwiping={(_, deltaX): void => onDrag(deltaX)}
                onSwiped={onDragEnd}
            >
                <BaseCarousel
                    selectedIndex={selectedIndex + dragIndexOffset}
                    visibleCount={visibleCount}
                    spacing={spacing}
                    animationDuration={isDragging ? 0 : undefined}
                >
                    {children}
                </BaseCarousel>
            </Swipeable>
        </div>
    );
}
