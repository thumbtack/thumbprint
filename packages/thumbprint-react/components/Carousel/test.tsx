import React from 'react';
import { mount } from 'enzyme';
import { noop } from 'lodash';
import Carousel from './index';

describe('Carousel', () => {
    test('renders a carousel', () => {
        const wrapper = mount(
            <Carousel selectedIndex={0} onSelectedIndexChange={noop}>
                gooose
            </Carousel>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('does not render duplicate children when not animating', () => {
        const wrapper = mount(
            <Carousel selectedIndex={0} onSelectedIndexChange={noop}>
                <div data-test="item">Goose</div>
            </Carousel>,
        );

        expect(wrapper.find('[data-test="item"]')).toHaveLength(1);
    });

    describe('`selectedIndex`', () => {
        test('supports decimal `selectedIndex` values', () => {
            const wrapper = mount(
                <Carousel selectedIndex={2.3} onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('supports negative `selectedIndex`', () => {
            const wrapper = mount(
                <Carousel selectedIndex={-0.8} onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('`visibleCount`', () => {
        test('one visible item', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} visibleCount={1} onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('one and a half visible items', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} visibleCount={1.5} onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('three visible items', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} visibleCount={3} onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('`spacing`', () => {
        test('no spacing', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} spacing="0px" onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('spacing in pixels', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} spacing="10px" onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        test('spacing in ems', () => {
            const wrapper = mount(
                <Carousel selectedIndex={0} spacing="2em" onSelectedIndexChange={noop}>
                    <div>1</div>
                    <div>2</div>
                    <div>3</div>
                    <div>4</div>
                </Carousel>,
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
