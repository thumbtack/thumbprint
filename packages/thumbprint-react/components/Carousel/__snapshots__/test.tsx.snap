// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel \`selectedIndex\` supports decimal \`selectedIndex\` values 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={2.3}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={2.3}
          spacing="0px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(-29.999999999999982%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`selectedIndex\` supports negative \`selectedIndex\` 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={-0.8}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={-0.8}
          spacing="0px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(-19.999999999999996%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`spacing\` no spacing 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  spacing="0px"
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="0px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`spacing\` spacing in ems 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  spacing="2em"
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="2em"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 2em)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "2em",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "2em",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "2em",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "2em",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`spacing\` spacing in pixels 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  spacing="10px"
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="10px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 10px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "10px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "10px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "10px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "10px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`visibleCount\` one and a half visible items 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  visibleCount={1.5}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="0px"
          visibleCount={1.5}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "66.66666666666666%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "66.66666666666666%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "66.66666666666666%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "66.66666666666666%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`visibleCount\` one visible item 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  visibleCount={1}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="0px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel \`visibleCount\` three visible items 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
  visibleCount={3}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="0px"
          visibleCount={3}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "33.33333333333333%",
                  }
                }
              >
                <div>
                  1
                </div>
              </li>
              <li
                className="item"
                key="1/.1"
                style={
                  Object {
                    "order": 1,
                    "paddingRight": "0px",
                    "width": "33.33333333333333%",
                  }
                }
              >
                <div>
                  2
                </div>
              </li>
              <li
                className="item"
                key="2/.2"
                style={
                  Object {
                    "order": 2,
                    "paddingRight": "0px",
                    "width": "33.33333333333333%",
                  }
                }
              >
                <div>
                  3
                </div>
              </li>
              <li
                className="item"
                key="3/.3"
                style={
                  Object {
                    "order": 3,
                    "paddingRight": "0px",
                    "width": "33.33333333333333%",
                  }
                }
              >
                <div>
                  4
                </div>
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;

exports[`Carousel renders a carousel 1`] = `
<Carousel
  onSelectedIndexChange={[Function]}
  selectedIndex={0}
>
  <div>
    <Swipeable
      delta={10}
      disabled={false}
      flickThreshold={0.6}
      nodeName="div"
      onSwiped={[Function]}
      onSwiping={[Function]}
      preventDefaultTouchmoveEvent={true}
      rotationAngle={0}
      stopPropagation={false}
    >
      <div
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
      >
        <BaseCarousel
          selectedIndex={0}
          spacing="0px"
          visibleCount={1}
        >
          <div
            className="root"
          >
            <ul
              className="wrapper"
              style={
                Object {
                  "transform": "translateX(0%)",
                  "transition": "transform 400ms ease",
                  "width": "calc(100% + 0px)",
                }
              }
            >
              <li
                className="item"
                key="0/.0"
                style={
                  Object {
                    "order": 0,
                    "paddingRight": "0px",
                    "width": "100%",
                  }
                }
              >
                gooose
              </li>
            </ul>
          </div>
        </BaseCarousel>
      </div>
    </Swipeable>
  </div>
</Carousel>
`;
